packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -191,15 +192,39 @@ abstract class WidgetsBinding extends BindingBase implements GestureBinding, Ren
   }
 
   void _handleBuildScheduled() {
-    // If we're in the process of building dirty elements, we're know that any
-    // builds that are scheduled will be run this frame, which means we don't
-    // need to schedule another frame.
-    if (_buildingDirtyElements)
-      return;
+    // If we're in the process of building dirty elements, then changes
+    // should not trigger a new frame.
+    assert(() {
+      if (debugBuildingDirtyElements) {
+        throw new FlutterError(
+          'Build scheduled during frame.\n'
+          'While the widget tree was being built, laid out, and painted, '
+          'a new frame was scheduled to rebuild the widget tree. '
+          'If a change is needed to the widget tree, it should be applied '
+          'as the tree is being built. Scheduling a change for the subsequent '
```

Comment:
setState during build will hit another assert (we ignore it if it's the same widget, we queue it if it's a descendant, we assert elsewhere if it's an ancestor or cousin). But I've added "This might be because setState() was called from a layout or paint callback", which is the usual problem if you hit this assert.